transmogrifier/sources/transformer.py

```diff
@@ -348,89 +364,10 @@ def _write_deleted_records_to_txt_file(
-                pa.field("run_type", pa.string()),
-                pa.field("run_id", pa.string()),
-                pa.field("action", pa.string()),
-            )
-        )
-        partition_columns = ["source", "run_date", "run_type", "action", "run_id"]
-
-        written_files = []
-        ds.write_dataset(
-            self.yield_record_rows_for_writing(),
-            schema=schema,
-            base_dir=dataset_location,
-            partitioning=partition_columns,
-            partitioning_flavor="hive",
-            format="parquet",
-            basename_template="records-{i}.parquet",
-            existing_data_behavior="delete_matching",
-            use_threads=True,
-            max_rows_per_group=PARQUET_DATASET_BATCH_SIZE,
-            max_rows_per_file=100_000,
-            file_visitor=lambda written_file: written_files.append(written_file),
-        )
-        return written_files
-
-    def yield_record_rows_for_writing(
-        self, batch_size: int = PARQUET_DATASET_BATCH_SIZE
-    ) -> Iterator:
-        """Prepare batches of transformed records for writing.
-
-        Each row in the batch includes a serialized form of the original source record,
-        the transformed TIMDEX record, and metadata from the input file used for
-        partitioning during write.
-
-        NOTE: this method is a WIP.  While functional, and demonstrative of v1 / v2
-        branching, it will likely undergo revisions and still needs testing.
-        """
-        records = []
-        for etl_record in self:
-
-            if not isinstance(etl_record, ETLRecord):
-                message = "ETLRecord instance required for batch creation"
-                raise TypeError(message)
-            if not isinstance(self.partition_values, dict):
-                message = "Transformer.partition_values is not a dictionary"
-                raise TypeError(message)
-
-            record = {
-                "timdex_record_id": etl_record.timdex_record_id,
-                "source_record": etl_record.serialized_source_record,
-                "transformed_record": etl_record.serialized_transformed_record(),
-                "source": self.partition_values["source"],
-                "run_date": self.partition_values["run_date"],
-                "run_type": self.partition_values["run_type"],
-                "run_id": self.run_id,
-                "action": etl_record.action,
-            }
-            records.append(record)
-
-            if len(records) >= batch_size:
-                batch = pa.RecordBatch.from_pylist(records)
-                yield batch
-                records = []
-
-        if records:
-            batch = pa.RecordBatch.from_pylist(records)
-            yield batch
+    def write_to_parquet_dataset(self, dataset_location: str) -> list:
+        """Write output to TIMDEX dataset."""
+        timdex_dataset = TIMDEXDataset(location=dataset_location)
+        return timdex_dataset.write(records_iter=self)
```

Comment:
Cutting is delivering! 🎉  This really validates the new library

Reply:
Agreed!

One concern I have is that we're outsourcing too much logic to this library, and Transmog becomes heavily dependent on it, versus just a "pipe" that it can use to output records (almost like `smart_open` where it's just a way to output objects handled by the application at hand).

I think _this_ writing part makes good sense, and is clean.  I'm a little more concerned that the `__next__()` method requires importing and yielding `timdex_dataset.api.DatasetRecord` instances... but it feels pretty necessary.

As a thought experiment: I _don't_ think it would be an unsual for an application that is going to write directly to parquet files to import things like `RecordBatch` or `Table` from `pyarrow` during writing.  So I think it's arguable that importing these from our library is analogous.

Reply:
To add, I do like how yielding `timdex_dataset.api.DatasetRecord` furthers the awareness that the `transmogrifier` and `timdex-dataset-api` are specifically built for TIMDEX. 🤔